homeassistant/components/telegram_bot/__init__.py

```diff
@@ -551,6 +605,149 @@ def _deprecate_timeout(hass: HomeAssistant, service: ServiceCall) -> None:
     )
 
 
+def _build_targets(
+    hass: HomeAssistant, service: ServiceCall
+) -> list[tuple[TelegramBotConfigEntry, int, str | None]]:
+    """Build list of targets where each target is represented by its corresponding config entry, chat ID and notify entity id (if target is a notify entity)."""
+
+    migrate_chat_ids = _warn_chat_id_migration(hass, service)
+
+    targets: list[tuple[TelegramBotConfigEntry, int, str | None]] = []
+
+    # build target list from notify entities using service data: `entity_id`
+
+    referenced = async_extract_referenced_entity_ids(
+        hass, TargetSelection(service.data)
+    )
+    notify_entity_ids = referenced.referenced | referenced.indirectly_referenced
+
+    # parse entity IDs
+    entity_registry = er.async_get(hass)
+    for notify_entity_id in notify_entity_ids:
+        # get config entry from notify entity
+        registry_entry = entity_registry.async_get(notify_entity_id)
+        if not registry_entry:
+            raise ServiceValidationError(
+                translation_domain=DOMAIN,
+                translation_key="action_failed",
```

Comment:
We should only use generic translations when we have to and we don't know any better. Here we know what the problem is and should say that in the translation.